added ubuntu-edu packages

d78ui98 opened the pull request:
added ubuntu-edu packages on buster branch. These packages are compiled in debian `buster`

closes #3 #7

d78ui98 committed:
added synced folder

d78ui98 committed:
added deb files

these .deb files are complied in debian buster from the source

d78ui98 committed:
modified provisioner script to install deb files

d78ui98 commented:
Instead of those 4 lines I can also do 
```
dpkg -i /vagrant/ubuntu-edu-*
```
this also works. Let me know if I should do this.

dashohoxha reviewed (commented):
No, this is wrong. This is not what I asked. This is unacceptable.

You should add the source code of the debian packages on the directory `deb/`. Then from those sources generate binary packages and install them (do this inside the installation script).

d78ui98 committed:
install.sh generates binary package and install

d78ui98 committed:
removed precompiled binary

d78ui98 committed:
added source package

d78ui98 committed:
removed binary package

d78ui98 committed:
added source folder

d78ui98 committed:
added edu.sh script

d78ui98 committed:
using default synced folder

d78ui98 committed:
added source edu.sh

akash0x53 commented on deps/edu.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
give file "executable" permission.

dashohoxha replied:
Akash, this file is sourced (included) from `install.sh`, so it is Ok even without "executable" permission. The same for root user check.

akash0x53 commented on deps/edu.sh:
```diff
@@ -0,0 +1,13 @@
+apt --yes install build-essential fakeroot devscripts equivs
+
+cd /vagrant/deps/source_package
+echo y | mk-build-deps -i debian/control
```
If root permissions not given then this script will fail, so its good to add root user check

d78ui98 replied:
akash vagrant provision script or script that it sources always run with root user.
I also checked it out once by adding `id` in `edu.sh`

akash0x53 commented on deps/source_package/debian/control:
```diff
@@ -0,0 +1,64 @@
+ 
+Package: debian-edu-preschool
+Architecture: any
+Depends: calibre, chemtool, fritzing, inkscape, kalzium, kmplot, ktouch, kturtle, lightspeed, marble, melting, pencil2d, rocs, step, yorick, ${germinate:Depends}, ${misc:Depends}
```
This package should not be in "depends" section, they should be in Recommends section.

d78ui98 replied:
Yes. But there is a little problem.  One cannot install recommended and dependent packages with `dpkg`. So we have to use `apt install -f` to fix dependencies.

We install recommended packages with flag `--install-recommends`. But combining the 2 will not work. I am trying to solve this one.

akash0x53 replied:
No. If you publish your deb file and you used `apt install debian-edu-*` it will install recommended package too. - `apt-get` does so. Why don't you setup PPA and upload these deb files and try to install it?

d78ui98 replied:
I was thinking of this. But I saw this at https://wiki.debian.org/Packaging/Intro - 
```
QUESTION: is there anything like PPA in Ubuntu?
ANSWER: Debian does not run a PPA service like Ubuntu, but all the tools to make apt repositories yourself are there, it's just a lot to configure by oneself.
```
But still we can do this. Since there is no other way of installing recommended packages other than  `apt`

dashohoxha replied:
> If you publish your deb file and you used apt install debian-edu-* it will install recommended package too.

The command `apt install ...` works with deb files too. No need for a PPA or something similar.

> However debian has mentors.debian.net I am uploading my packages there

First install them internally (what you have done so far). Uploading them to debian should be the next step, because it is more complex and requires more responsibility. You don't just upload them to debian but you promise to maintain them for a long time, even after the GSoC project is finished, and you take responsibility for them.

As I have told you previously, it is better to find some EDU related project on debian, and submit them to this project (or through this project, I don't know how it works), rather than submitting them directly to debian.

dashohoxha replied:
There is a problem with `apt install debian-edu-*.deb`, that we don't give the user a choice about which packages to install.
See: https://github.com/docker-scripts/dev--LTSP/issues/59

akash0x53 commented on deps/source_package/desktop-gnome-amd64:
```diff
@@ -0,0 +1 @@
+gnome-session-flashback
```
what is the purpose of this file?

d78ui98 replied:
removed. Along with other unnecessary files.

akash0x53 commented on install.sh:
```diff
@@ -9,7 +9,10 @@ apt --yes --install-recommends install dnsmasq
 apt --yes --install-recommends install ltsp-server
 DEBIAN_FRONTEND=noninteractive apt --yes --install-recommends install ltsp-client
 apt --yes install epoptes epoptes-client
+apt --yes install build-essential fakeroot devscripts equivs
```
edu.sh also updating apt repo and installing these packages.

akash0x53 commented on deps/edu.sh:
```diff
@@ -0,0 +1,13 @@
+cd /vagrant/deps/source_package
+echo y | mk-build-deps -i debian/control
+dpkg-buildpackage -us -uc
+dpkg -i ../debian-edu-preschool_15.12.5_amd64.deb
```
can you try dpkg -i ../debian-edu-*.deb ?

d78ui98 replied:
yes.

dashohoxha commented:
Deepanshu, you have added about 60 new files. I understand that you have just copied the source files of the edubuntu package, since this is easier for you.

Although this may work, it seems to me a lot of files. I would prefer something simpler, for example having only these 4 directories: `deb/debian-edu-preschool/`, `deb/debian-edu-primary`, `deb/debian-edu-secondary`, `deb/debian-edu-terciary/`, where each of these directories has about 4-5 control files.

This may require you to learn how to build deb packages, to understand the structure of control files etc. But that is the point, I want you to learn and to understand deb packages.

Since I don't understand deb packages myself, I am not sure that what I am asking is possible. But I believe there should be an easier way, you don't need 60 files just to build 4 packages.

Maybe you should ask on the mailing list for any good starting tutorial about understanding and building deb packages.

d78ui98 committed:
removed packages installed twice

d78ui98 committed:
updated edu.sh

added comments, combined dpkg commands into one

d78ui98 committed:
removed unnecessary lines from control file

d78ui98 committed:
removed unnecessary files

d78ui98 committed:
updated metapackage-map

d78ui98 commented:
>I am not sure that what I am asking is possible.

Yes we can do this. I just have to create write 4 different `control` file in 4 different folders(currently all are defined in one). H

dashohoxha commented:
> However, it would increase number of files

I am in favor of a minimalist solution, using the least number of files possible, the least configuration fields possible, the least tools possible. For example use no tools other that **dpkg-deb** for building the debian packages.

akash0x53 commented:
+1 for dpkg-deb

d78ui98 commented:
>I am in favor of a minimalist solution

Agreed. dpkg-deb comes from `devscripts` package. Other than that we only need `equivs` to fix the dependencies. Otherwise `mk-build-deps` will not work.

d78ui98 committed:
removed unecessary packages

d78ui98 committed:
removed optional files

d78ui98 commented:
We are now down to 6 configuration files. 5 are inside debian folder. Any of them cannot be removed.
I refered to this recommended guide from mailing list - https://wiki.debian.org/Packaging/Intro

dashohoxha commented:
I suggest that you use this Docker container for testing these scripts and config files: https://github.com/docker-scripts/debian
But you have to replace `stretch` with `buster` on the base image:
https://github.com/docker-scripts/debian/blob/master/Dockerfile#L1

d78ui98 commented:
Sure.

akash0x53 commented on deps/source_package/update:
```diff
@@ -1,13 +0,0 @@
-#! /bin/sh
-
-if ! which dch >/dev/null; then
```
Replaced with `hash` - check bash builtins.

akash0x53 commented on deps/source_package/update:
```diff
@@ -1,13 +0,0 @@
-fi
-
-if ! which debootstrap >/dev/null; then
-	echo >&2 "please install debootstrap"
```
you can use `for` loop here instead of using 2 `if` statements.

d78ui98 replied:
This file has been removed. Since it is optional.

d78ui98 commented:
> I suggest that you use this Docker container for testing these scripts and config files

yes they work perfectly.

dashohoxha commented:
> yes they work perfectly.

Did you use the 'debian' container for testing the packages?
What programs are installed when you install the package `debian-edu-preschool`?

dashohoxha commented:
> We are now down to 6 configuration files. 5 are inside debian folder. Any of them cannot be removed.

But their content can be fixed and improved. For example:
- `changelog` is too long; delete the previous content and write your own change
- `compat` maybe should be higher than 7
- `control` should be fixed; remove `germinate` and fix the list of dependencies for each package
- `copyright` should be fixed (or maybe deleted)
- remove `germinate` from `rules`
- `metapackage-map` maybe should be removed
- etc.

> I refered to this recommended guide from mailing list - https://wiki.debian.org/Packaging/Intro

This is a good one. But the ones that I suggested are good as well: https://github.com/docker-scripts/dev--LTSP/issues/7#issuecomment-400156930

akash0x53 commented on deps/source_package/debian/compat:
```diff
@@ -0,0 +1 @@
+7
```
You can set this to 10

https://www.debian.org/doc/manuals/maint-guide/dother.en.html#compat

akash0x53 commented on deps/source_package/debian/changelog:
```diff
@@ -0,0 +1,4570 @@
+edubuntu-meta (15.12.5) xenial; urgency=medium
```
We can keep only this year's changelog

d78ui98 replied:
Yes. I am keeping some lines of `changelog`

dashohoxha replied:
I don't think we need any previous logs, since this is not a continuation of ubuntu-meta.
Just the initial log of Deepanshu is enough (and it should mention debian-edu instead of ubuntu-edu).

Anyway, since this is just an internal package, that has no chance to be published on DebianEdu, it doesn't matter much what is written on the log. After this project, maybe this package will not be maintained. It served the purpose of helping us to learn how to build deb packages, and in my opinion this is fine.

d78ui98 replied:
sure. will create new `changelog` with `dch`

d78ui98 committed:
updated changelog

d78ui98 committed:
removed germinate

d78ui98 committed:
removed metapackage-map

d78ui98 committed:
updated rules removed copyright

dashohoxha commented on deps/source_package/debian/control:
```diff
@@ -0,0 +1,35 @@
+Depends: calibre, cantor, chemtool, fritzing, inkscape, kalzium, kmplot, ktouch, kturtle, lightspeed, marble, melting,pencil2d, rocs, step, yorick
+Description: Tertiary Educational Application Bundle
+ This package depends on all of the educational software for Tertiary grade
+ level education that is fully supported.
```
You still need to fix the list of dependences for each package. For example the dependences of `debian-edu-preschool` are the same as those of `debian-edu-terciary`. This cannot be right. Why are you so careless?

akash0x53 replied:
@d78ui98 compare dependencies here,

preschool: https://packages.ubuntu.com/bionic/ubuntu-edu-preschool
tertiary: https://packages.ubuntu.com/bionic/ubuntu-edu-tertiary

d78ui98 replied:
Thanks.

d78ui98 committed:
fixed dependencies of debian-edu-preschool

d78ui98 committed:
renamed source folder

it is recommended to have source folder in the format of source-name then version

d78ui98 committed:
removed extra lines from changelogs

d78ui98 committed:
changed source to edudebian

d78ui98 committed:
updated edu.sh accordingly

dashohoxha commented on deps/edu.sh:
```diff
@@ -0,0 +1,13 @@
+apt --yes install devscripts equivs
+
+# Building debian-edu packages
+cd /vagrant/deps/edudebian-meta-1.0
```
Is this version number needed? When you make a new version of the package you don't make a new directory, do you?
What does this `deps` mean? I think it should be `packages` or `deb`.

akash0x53 replied:
`deps` can be renamed to something like `edu-packages`.

d78ui98 replied:
>Is this version number needed?

No, it will work without it as well. Just give us a warning.
```
dpkg-source: warning: source directory 'edudebian-meta' is not <sourcepackage>-<upstreamversion> 'edudebian-meta-1.0'
```
But as we are usiing these packages for our own purposes we can do that.

d78ui98 committed:
renamed to edu-packages

d78ui98 committed:
renamed to edudebian-meta

d78ui98 committed:
source path changed

d78ui98 committed:
updated source path in install.sh
